src/benchmarklib/benchmark_runner.cpp

```diff
@@ -0,0 +1,449 @@
+        const auto file_prefix = name + "-LQP-" + std::to_string(lqp_idx);
+        LQPVisualizer{graphviz_config, {}, {}, {}}.visualize({lqps[lqp_idx]}, file_prefix + ".dot",
+                                                             file_prefix + ".svg");
+      }
+      for (auto pqp_idx = size_t{0}; pqp_idx < pqps.size(); ++pqp_idx) {
+        const auto file_prefix = name + "-PQP-" + std::to_string(pqp_idx);
+        SQLQueryPlanVisualizer{graphviz_config, {}, {}, {}}.visualize(*pqps[pqp_idx], file_prefix + ".dot",
+                                                                      file_prefix + ".svg");
+      }
+    }
+  }
+}
+
+void BenchmarkRunner::_benchmark_permuted_query_sets() {
+  // Init results
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    _query_results_by_query_name.emplace(name, QueryBenchmarkResult{});
+  }
+
+  auto mutable_named_queries = _queries;
+
+  // For shuffling the query order
+  std::random_device random_device;
+  std::mt19937 random_generator(random_device());
+
+  BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
+  while (state.keep_running()) {
+    std::shuffle(mutable_named_queries.begin(), mutable_named_queries.end(), random_generator);
+
+    for (const auto& named_query : mutable_named_queries) {
+      const auto query_benchmark_begin = std::chrono::steady_clock::now();
+
+      // Execute the query, we don't care about the results
+      _execute_query(named_query);
+
+      const auto query_benchmark_end = std::chrono::steady_clock::now();
+
+      auto& query_benchmark_result = _query_results_by_query_name.at(named_query.first);
+      query_benchmark_result.duration += query_benchmark_end - query_benchmark_begin;
+      query_benchmark_result.num_iterations++;
+    }
+  }
+}
+
+void BenchmarkRunner::_benchmark_individual_queries() {
+  for (const auto& named_query : _queries) {
+    const auto& name = named_query.first;
+    _config.out << "- Benchmarking Query " << name << std::endl;
+
+    BenchmarkState state{_config.max_num_query_runs, _config.max_duration};
+    while (state.keep_running()) {
+      _execute_query(named_query);
+    }
+
+    QueryBenchmarkResult result;
+    result.num_iterations = state.num_iterations;
+    result.duration = std::chrono::high_resolution_clock::now() - state.begin;
+
+    _query_results_by_query_name.emplace(name, result);
+  }
+}
+void BenchmarkRunner::_execute_query(const NamedQuery& named_query) {
+  const auto& name = named_query.first;
+  const auto& sql = named_query.second;
+
+  auto pipeline = SQLPipelineBuilder{sql}.with_mvcc(_config.use_mvcc).create_pipeline();
+  // Execute the query, we don't care about the results
+  pipeline.get_result_table();
+
+  // If necessary, keep plans for visualization
+  if (_config.enable_visualization) {
+    const auto query_plans_iter = _query_plans.find(name);
+    if (query_plans_iter == _query_plans.end()) {
```

Comment:
I don't think that we should do this lookup for every iteration.

Reply:
would you suggest just storing the plans in every iteration instead of checking whether the exist already?

Reply:
I thought we should take it out of the main loop, but thinking about it again, we already have the `enable_visualization` check. If that is on, we don't care about the performance of the lookup too much